### ISSUE TITLE
Rework 'revoke all tokens' flow

### ISSUE THREAD

**Issue by monkeytypegeorge** [monkeytypegame/monkeytype#4684]
Currently, due to the token cache, it is not working as expected

**Comment by aikooo7** [comment]
Hello @Miodec / @monkeytypegeorge , I want to help so you want to change the flow to what?

**Comment by Warrobot10** [comment]
> Hello @Miodec / @monkeytypegeorge , I want to help so you want to change the flow to what?

At the v1.17.11 release (we are now using a different versioning system) the revoke all tokens button was added to dangerzone in settings. This basically signs people out of their accounts on all of their devices, which is useful if they believe their account has been compromised.

![image](https://github.com/monkeytypegame/monkeytype/assets/132307750/abbabe12-e264-4237-a33d-154fbaf08769)
 
I recommend you check the [commit](https://github.com/monkeytypegame/monkeytype/commit/87e882b94de8d0a0a1d387ef1ccf223be30b0dfc) so you can get a sense of what is happening and how the code works.

**Comment by aikooo7** [comment]
Hello @Warrobot10 appreciate your help and understand what it does now, so I want to know what you want it to be changed? Does it not work properly and that is what needs to be fixed?

**Comment by Warrobot10** [comment]
I am not sure what @Miodec is thinking of changing, but I can provide a suggestion. Currently, when the user wants to revoke all tokens, they need to input their password. This is inherently flawed, because the reason they are trying to sign out all sessions is because they believe the account is compromised. This means the attacker knows the password to that account, and can change the password because they know the password. Now that they have changed the password, the account owner cannot revoke all tokens, thus defeating the purpose of the revoke all tokens. The account is essentially hijacked, and the account owner locked out as soon as the attacker has the password. The attacker could also use the revoke all tokens against the account owner, logging them out once and for all.

**Comment by aikooo7** [comment]
Thanks again @Warrobot10 , we are in double faced knife here.

Because is really hard to do that without password and without it I think it would make it even worse.

Anyways, if that is the wish of miodec I will do it for sure since is his project after all, thanks again warrobot.

Regards,
aikoo7.

**Comment by Miodec** [comment]
Requiring the password is fine, im pretty sure thats how it should be done.

The issue is about the current implementation of token cache. Because decoding the firebase token takes a while, we've implemented a token cache which shortens the time to authenticate from hundredths of miliseconds to just a couple. 

But, this means that the tokens will be cached until their expiration date, and WON'T be checked if they have been revoked.

You can test it by signing in on two separate browsers, then revoking tokens on one of them - the other browser, while signed in, will still be able to make changes to the account, because their auth token is cached.

**Comment by Miodec** [comment]
The first thing that comes to mind is keeping a list of uid's that revoked tokens and the timestamp of when they did so. Then every time we auth users, we check if they are in that list - if they arent, they are authed. If they are, we check if their token creation date is after the revoke date.

**Comment by aikooo7** [comment]
@Miodec I liked your idea but here is a suggestion, why not add a field in a db column like "revoked" and when a user logs in, you check the email, password, and it?

**Comment by Hattorius** [comment]
Yes, as @Miodec and, essentially, @aikooo7 have suggested, it's advisable to maintain a record of token revocation status. However, the challenge lies in the fact that this approach introduces an added cost to every authenticated API request. Each request necessitates establishing a database connection to ascertain whether the token should be revoked or not. While it's true that this increases the resource consumption per request, it's important to emphasize that this measure is primarily intended for enhancing security, especially in cases where an account may have been compromised.

Consider, for instance, the 'loading user data' phase that occurs when a user logs in and the page is initially loaded. At this juncture, a GET request is made to the `/users` endpoint. Implementing token revocation checks at this stage can be relatively straightforward, particularly because the user data has already been retrieved from the database, making it a simple if statement. However, it's essential to acknowledge that this alone may not provide an impenetrable defense against determined attackers who may still attempt to access other endpoints.

**Comment by Hattorius** [comment]
> The first thing that comes to mind is keeping a list of uid's that revoked tokens and the timestamp of when they did so. Then every time we auth users, we check if they are in that list - if they arent, they are authed. If they are, we check if their token creation date is after the revoke date.

I think this would be the better option. Would updating the `User` interface with a new column suffice? Was thinking about a column that just holds the last time all tokens were revoked: this way you only need to check if the token's issue date is later than the date shown in the column.

If this is an acceptable solution, can I get this issue assigned to me? @Miodec

**Comment by Miodec** [comment]
> > The first thing that comes to mind is keeping a list of uid's that revoked tokens and the timestamp of when they did so. Then every time we auth users, we check if they are in that list - if they arent, they are authed. If they are, we check if their token creation date is after the revoke date.
> 
> I think this would be the better option. Would updating the `User` interface with a new column suffice? Was thinking about a column that just holds the last time all tokens were revoked: this way you only need to check if the token's issue date is later than the date shown in the column.
> 
> If this is an acceptable solution, can I get this issue assigned to me? @Miodec

Im still a bit worried about a DB lookup on EVERY authentication attempt (which at this point would be like 30 per second).

If we go for this we need to do some performance testing to estimate the impact.

Assigning you @Hattorius

**Comment by Hattorius** [comment]
> > > The first thing that comes to mind is keeping a list of uid's that revoked tokens and the timestamp of when they did so. Then every time we auth users, we check if they are in that list - if they arent, they are authed. If they are, we check if their token creation date is after the revoke date.
> > 
> > 
> > I think this would be the better option. Would updating the `User` interface with a new column suffice? Was thinking about a column that just holds the last time all tokens were revoked: this way you only need to check if the token's issue date is later than the date shown in the column.
> > If this is an acceptable solution, can I get this issue assigned to me? @Miodec
> 
> Im still a bit worried about a DB lookup on EVERY authentication attempt (which at this point would be like 30 per second).
> 
> If we go for this we need to do some performance testing to estimate the impact.
> 
> Assigning you @Hattorius

I completely agree with your line of thinking here. It's challenging to come up with a more reliable solution. One idea I had was to enhance the security by incorporating the user ID and timestamp of revocation into Redis. By doing this, we could perform a quick preliminary check on the user ID and timestamp data from the JWT token before even verifying the token's validity. This approach, however, might impose a greater load on the Redis server, akin to how Redis AOF operates.

Alternatively, we'd be looking at the additional 30 requests per second to the MongoDB server. It's crucial to emphasize that user security is paramount in this context, so any additional overhead on the Redis server should be carefully considered.

Should I proceed with implementing the MongoDB check first to assess the impact on server load? @Miodec

**Comment by Miodec** [comment]
Yeah, let's check the impact on mongo first

**Comment by Hattorius** [comment]
I've made an addition to the `User` interface, which involves a new field that captures the current timestamp when the tokens are revoked. The authentication process on API endpoints now includes an initial check against the database to determine if a token should be revoked, and everything seems to be functioning as intended.

However, there's a specific scenario where our approach needs refinement. Currently, password management is outsourced to Google's Firebase, and updates to a user's password bypass Monkeytype's backend entirely. Consequently, this leads to a situation where the token can still be used to reset the password.

To address this issue, I propose the implementation of a new API endpoint. This endpoint would enable us to validate both the password and token on our side before forwarding the password change request to Firebase on the server side.

I'd appreciate any thoughts on this approach. If you have any alternative solutions in mind of if you agree with the proposal, please let me know.

**Comment by Miodec** [comment]
I think, since its going directly to the Firebase API, revoke all tokens should work immediately there, because its not going through our token cache. Have you checked if its still possible to update the password even after revoking?

**Comment by Hattorius** [comment]
> I think, since its going directly to the Firebase API, revoke all tokens should work immediately there, because its not going through our token cache. Have you checked if its still possible to update the password even after revoking?

I share your initial assumption that revoking all tokens should effectively block all actions, as it directly affects the Firebase API without going through our token cache. However, I encountered an unexpected behavior. When I tested this scenario using a different browser with a distinct token while logged into the same account, I observed that even after revoking all session tokens, it was still possible to change the password on the other session.

This behavior has left me puzzled, as all other actions were correctly blocked. To illustrate the issue, I've recorded a video that demonstrates the situation. Could you please take a moment to watch the video and let me know if I might be missing something or if there's a potential solution? https://www.youtube.com/watch?v=Z7cAmUAz-Aw

**Comment by Miodec** [comment]
Man, firebase continues to amaze me with its lack of / broken features. I keep thinking about just making a custom auth system, but keep getting scared by the weight of this change. Eh.

Anyway, yeah, we gonna need to move all the client side stuff (so, linking a new credential, unlinking, updating password) to the api server and add an endpoint - just like we already do with `updateEmail`.

If you would like to work on this, please split this into 2 prs - one for adding the endpoints and one for the actual token revoke status change.

**Comment by Hattorius** [comment]
I had the same idea of creating some kind of SSO service that's 100% open source, but I was still brainstorming ideas. It's kind of surprising, to be honest. I would've thought **Google's** Firebase would be a solid option, but you *shouldn't* be able to change the password with a revoked token.

I'm totally up for moving the client-side password change to the backend, and I'll split that into two separate pull requests. But, I'm wondering, does this really solve the issue? Now that we're using still Firebase, which seems a bit insecure (it ran on the frontend, and anyone can initiate Firebase to change a password since all the data is available on the frontend), is there a way to set up an extra security rule within Firebase to prevent these requests (and only accept requests from the backend)?

**Comment by Miodec** [comment]
Im not sure if i can selectively block features like that..

**Comment by Hattorius** [comment]
I'm pretty sure it's possible using custom claims. See:
 - [Custom-claim attributes and roles](https://firebase.google.com/docs/rules/basics#custom-claim_attributes_and_roles)
 - [Custom claims](https://firebase.google.com/docs/auth/admin/custom-claims)

**Comment by Miodec** [comment]
Do you think you could join the Discord server (discord.gg/monkeytype) and message in the #development channel so we can discuss it in real time?
